go/signedexchange/signedexchange.go

```diff
@@ -0,0 +1,187 @@
+package signedexchange
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/WICG/webpackage/go/signedexchange/cbor"
+	"github.com/WICG/webpackage/go/signedexchange/mice"
+)
+
+type Input struct {
+	// Request
+	requestUri *url.URL
+
+	// Response
+	responseStatus int
+	responseHeader http.Header
+
+	// Payload
+	payload []byte
+}
+
+func NewInput(uri *url.URL, status int, headers http.Header, payload []byte, miRecordSize int) (*Input, error) {
+	i := &Input{
+		requestUri:     uri,
+		responseStatus: status,
+		responseHeader: headers,
+	}
+	if err := i.miEncode(payload, miRecordSize); err != nil {
+		return nil, err
+	}
+	return i, nil
+}
+
+func (i *Input) miEncode(payload []byte, recordSize int) error {
+	var buf bytes.Buffer
+	mi, err := mice.Encode(&buf, payload, recordSize)
+	if err != nil {
+		return err
+	}
+	i.payload = buf.Bytes()
+	i.responseHeader.Add("Content-Encoding", "mi-sha256")
+	i.responseHeader.Add("MI", mi)
+	return nil
+}
+
+// AddSignedHeadersHeader adds 'signed-headers' header to the response.
+//
+// Signed-Headers is a Structured Header as defined by
+// [I-D.ietf-httpbis-header-structure]. Its value MUST be a list (Section 4.8
+// of [I-D.ietf-httpbis-header-structure]) of lowercase strings (Section 4.2 of
+// [I-D.ietf-httpbis-header-structure]) naming HTTP response header fields.
+// Pseudo-header field names (Section 8.1.2.1 of [RFC7540]) MUST NOT appear in
+// this list.
+func (i *Input) AddSignedHeadersHeader(ks ...string) {
+	strs := []string{}
+	for _, k := range ks {
+		strs = append(strs, fmt.Sprintf(`"%s"`, strings.ToLower(k)))
+	}
+	s := strings.Join(strs, ", ")
+	i.responseHeader.Add("signed-headers", s)
+}
+
+func (i *Input) AddSignatureHeader(s *Signer) error {
+	h, err := s.signatureHeaderValue(i)
+	if err != nil {
+		return err
+	}
+	i.responseHeader.Add("Signature", h)
+	return nil
+}
+
+func (i *Input) parseSignedHeadersHeader() []string {
+	unparsed := i.responseHeader.Get("signed-headers")
+
+	rawks := strings.Split(unparsed, ",")
+	ks := make([]string, 0, len(rawks))
+	for _, k := range rawks {
+		ks = append(ks, strings.Trim(k, "\" "))
+	}
+	return ks
+}
+
+func (i *Input) encodeCanonicalRequest(e *cbor.Encoder) error {
+	mes := []*cbor.MapEntryEncoder{
+		cbor.GenerateMapEntry(func(keyE *cbor.Encoder, valueE *cbor.Encoder) {
+			keyE.EncodeByteString([]byte(":method"))
+			valueE.EncodeByteString([]byte("GET"))
+		}),
+		cbor.GenerateMapEntry(func(keyE *cbor.Encoder, valueE *cbor.Encoder) {
+			keyE.EncodeByteString([]byte(":url"))
+			valueE.EncodeByteString([]byte(i.requestUri.String()))
+		}),
+	}
+	return e.EncodeMap(mes)
+}
+
+func (i *Input) encodeResponseHeader(e *cbor.Encoder, onlySignedHeaders bool) error {
```

Comment:
Be consistent about whether the HTTP header fields are one "header" or a list of "headers". I tend to prefer making it plural, but one can argue that that's not literally correct.